pkg/capacity/capacity.go

```diff
@@ -228,24 +232,28 @@ var _ metrics.StableCollector = &Controller{}
 
 // Run is a main Controller handler
 func (c *Controller) Run(ctx context.Context, threadiness int) {
-	klog.Info("Starting Capacity Controller")
+	logger := c.logger
+
+	logger.Info("starting controller")
 	defer c.queue.ShutDown()
 
-	c.prepare(ctx)
+	c.prepare(ctx, logger)
 	for i := 0; i < threadiness; i++ {
 		go wait.UntilWithContext(ctx, func(ctx context.Context) {
-			c.runWorker(ctx)
+			c.runWorker(ctx, logger)
 		}, time.Second)
 	}
 
-	go wait.UntilWithContext(ctx, func(ctx context.Context) { c.pollCapacities() }, c.pollPeriod)
+	go wait.UntilWithContext(ctx, func(ctx context.Context) { c.pollCapacities(logger) }, c.pollPeriod)
 
-	klog.Info("Started Capacity Controller")
+	logger.Info("started controller")
 	<-ctx.Done()
-	klog.Info("Shutting down Capacity Controller")
+	logger.Info("shutting down controller")
 }
 
-func (c *Controller) prepare(ctx context.Context) {
+func (c *Controller) prepare(ctx context.Context, logger logr.Logger) {
+	logger = logger.WithName("prepare")
```

Comment:
I have mixed feelings about explicitly passing a logger here and in the other functions which already take a context.

One the one hand it makes it explicit that the function is going to use a `logr.Logger` and needs it.

On the other hand, it is a redundant parameter because `ctx` could be used to provide the logger. A function that retrieves the logger from the context and uses `klogr.New` as fallback, i.e. never returns nil, would be ideal.

At least here we *can* add a parameter. There are plenty of established APIs which take a context, but cannot be changed easily to also take a logger - see client-go.

This is one of the open questions around the usage of `logr` in Kubernetes...

Reply:
For comparison, in `capacity_test.go` I used `logr.FromContext` because that file controlled the context and could be sure to not get nil.